packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -0,0 +1,446 @@
+
+  /**
+   * Specify an optional className to be applied to the modal node
+   */
+  containerClassName?: string;
+
+  /**
+   * Specify whether the primary button should be replaced with danger button.
+   * Note that this prop is not applied if you render primary/danger button by yourself
+   */
+  danger?: boolean;
+
+  /**
+   * Specify whether the Modal content should have any inner padding.
+   */
+  isFullWidth?: boolean;
+
+  /**
+   * Specify an optional handler for closing modal.
+   * Returning `false` here prevents closing modal.
+   */
+  onClose?(event: MouseEvent): void | boolean;
+
+  /**
+   * Called for all `onKeyDown` events that do not close the modal
+   */
+  onKeyDown?(evt: KeyboardEvent): void;
+
+  /**
+   * Specify whether the Modal is currently open
+   */
+  open?: boolean;
+
+  preventCloseOnClickOutside?: boolean;
+
+  /**
+   * Specify a CSS selector that matches the DOM element that should be
+   * focused when the Modal opens
+   */
+  selectorPrimaryFocus?: string;
+
+  /** Specify the CSS selectors that match the floating menus. */
+  selectorsFloatingMenus?: Array<string | null | undefined>;
+
+  size?: 'xs' | 'sm' | 'md' | 'lg';
+}
+
+const ComposedModal = React.forwardRef<HTMLDivElement, ComposedModalProps>(
+  function ComposedModal(
+    {
+      ['aria-labelledby']: ariaLabelledBy,
+      ['aria-label']: ariaLabel,
+      children,
+      className: customClassName,
+      containerClassName,
+      danger,
+      isFullWidth,
+      onClose,
+      onKeyDown,
+      open,
+      preventCloseOnClickOutside,
+      selectorPrimaryFocus,
+      selectorsFloatingMenus,
+      size,
+      ...rest
+    },
+    ref
+  ) {
+    const prefix = usePrefix();
+    const [isOpen, setIsOpen] = useState<boolean>(!!open);
+    const [wasOpen, setWasOpen] = useState<boolean>(!!open);
+    const innerModal = useRef<HTMLDivElement>(null);
+    const button = useRef<HTMLButtonElement>(null);
+    const startSentinel = useRef<HTMLButtonElement>(null);
+    const endSentinel = useRef<HTMLButtonElement>(null);
+
+    // Kepp track of modal open/close state
+    // and propagate it to the document.body
+    useEffect(() => {
+      if (open !== wasOpen) {
+        setIsOpen(!!open);
+        setWasOpen(!!open);
+        toggleClass(document.body, `${prefix}--body--with-modal-open`, !!open);
+      }
+    }, [open, wasOpen, prefix]);
+    // Remove the document.body className on unmount
+    useEffect(() => {
+      return () => {
+        toggleClass(document.body, `${prefix}--body--with-modal-open`, false);
+      };
+    }, []); // eslint-disable-line react-hooks/exhaustive-deps
+
+    function handleKeyDown(evt: KeyboardEvent) {
+      if (match(evt, keys.Escape)) {
+        closeModal(evt);
+      }
+
+      onKeyDown?.(evt);
+    }
+    function handleMousedown(evt: MouseEvent) {
+      const isInside = innerModal.current?.contains(evt.target as Node);
+      if (!isInside && !preventCloseOnClickOutside) {
```

Comment:
@francinelucca here is the fix for modal not closing on outside clicks ... as a rant sidenote: I find frankly fascinating that the previous version ever worked… that was some quite creative nonsense :thinking: